src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioDocumentNavigationService.cs

```diff
@@ -304,48 +307,70 @@ static VsTextSpan GetVsTextSpan(SourceText text, int position, int virtualSpace)
                         // Otherwise, just pick one of the ids to use for navigation.
                         var documentIdToNavigate = documentIdsForFilePath.Contains(documentId) ? documentId : documentIdsForFilePath.First();
-                        return await NavigateToFileInWorkspaceAsync(
-                            documentIdToNavigate, workspace, getVsTextSpan, cancellationToken).ConfigureAwait(false);
+                        return GetNavigationCallback(documentIdToNavigate, workspace, getVsTextSpan);
                     }
 
-                    return await TryNavigateToMappedFileAsync(
+                    return await GetNavigableLocationForMappedFileAsync(
                         workspace, document, mappedSpan.Value, cancellationToken).ConfigureAwait(false);
                 }
             }
 
-            return await NavigateToFileInWorkspaceAsync(
-                documentId, workspace, getVsTextSpan, cancellationToken).ConfigureAwait(false);
+            return GetNavigationCallback(documentId, workspace, getVsTextSpan);
         }
 
-        private async Task<bool> NavigateToFileInWorkspaceAsync(
+        private Func<CancellationToken, Task<bool>>? GetNavigationCallback(
             DocumentId documentId,
             Workspace workspace,
-            Func<SourceText, VsTextSpan> getVsTextSpan,
-            CancellationToken cancellationToken)
+            Func<SourceText, VsTextSpan> getVsTextSpan)
         {
-            // Have to be on the UI thread to open a document.
-            await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
-            var document = OpenDocument(workspace, documentId);
-            if (document == null)
+            return async cancellationToken =>
             {
-                return false;
-            }
+                // Always open the document again, even if the document is already open in the 
+                // workspace. If a document is already open in a preview tab and it is opened again 
+                // in a permanent tab, this allows the document to transition to the new state.
+
+                if (workspace.CanOpenDocuments)
+                    await OpenDocumentAsync(_threadingContext, workspace, documentId, cancellationToken).ConfigureAwait(false);
 
-            // Keep this on the UI thread since we're about to do span mapping.
-            var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(true);
-            var textBuffer = text.Container.GetTextBuffer();
+                if (!workspace.IsDocumentOpen(documentId))
+                    return false;
+
+                // Now that we've opened the document reacquire the corresponding Document in the current solution.
+                var document = workspace.CurrentSolution.GetDocument(documentId);
+                if (document == null)
+                    return false;
```

Comment:
Hmm, this is something else that would have also been busted by https://github.com/dotnet/roslyn/pull/59211 since we're assuming that's happening. Nothing impactful for this PR; I wanted to tag it so when we look to resurrect #59211 we can remember to look at this.

Reply:
Yes.  I tlikely would be.  We can talk more offline, however i think what we likely want is an async OpenFileAsync that can be non-UI-thread affinitized, and which returns back the actual *opened document* that later code can use.

Reply:
Indeed, that would be useful here.